napari/types.py

```diff
@@ -76,16 +76,16 @@ class SampleDict(TypedDict):
 # while their names should not change (without deprecation), their typing
 # implementations may... or may be rolled over to napari/image-types
 
-ArrayBase = np.ndarray
+ArrayBase: Type[np.ndarray] = np.ndarray
 
 
-ImageData = NewType("ImageData", ArrayBase)
-LabelsData = NewType("LabelsData", ArrayBase)
-PointsData = NewType("PointsData", ArrayBase)
-ShapesData = NewType("ShapesData", List[ArrayBase])
-SurfaceData = NewType("SurfaceData", Tuple[ArrayBase, ArrayBase, ArrayBase])
-TracksData = NewType("TracksData", ArrayBase)
-VectorsData = NewType("VectorsData", ArrayBase)
+ImageData = NewType("ImageData", np.ndarray)
+LabelsData = NewType("LabelsData", np.ndarray)
+PointsData = NewType("PointsData", np.ndarray)
+ShapesData = NewType("ShapesData", List[np.ndarray])
+SurfaceData = NewType("SurfaceData", Tuple[np.ndarray, np.ndarray, np.ndarray])
+TracksData = NewType("TracksData", np.ndarray)
+VectorsData = NewType("VectorsData", np.ndarray)
```

Comment:
Why are these changed? It's incorrect to say that these data are NumPy arrays — we support many other types of arrays. The idea was to expand ArrayBase to those arrays in the future, rather than to hardcode np.ndarray.

Reply:
Here is mypy error
```
napari/types.py:82: error: Argument 2 to NewType(...) must be subclassable (got ArrayBase?)  [valid-newtype]
napari/types.py:82: error: Variable "napari.types.ArrayBase" is not valid as a type  [valid-type]
napari/types.py:82: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
```

`NewType` cannot be used for described scenario. From [documentation](https://docs.python.org/3/library/typing.html#newtype)

> In contrast, NewType declares one type to be a subtype of another

Did we would like to convert it to [ArrayLike](https://numpy.org/doc/stable/reference/typing.html#arraylike)

I could try to rewrite it, but I'm unsure which specification to specify.

Reply:
> Did we would like to convert it to [ArrayLike](https://numpy.org/doc/stable/reference/typing.html#arraylike)

I think I like this solution better than the current solution.

Reply:
Here error for ArrayLike
```
napari/types.py:82: error: Argument 2 to NewType(...) must be subclassable (got "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]")  [valid-newtype]
```

It looks like NewType requires contract class. We may try to use [TypeVar](https://docs.python.org/3/library/typing.html#typing.TypeVar)

What did you think?

Reply:
I'm out of my depth with typing but TypeVar could be good option...? I was never clear of the advantages/disadvantages between the two.

Reply:
The new type is way to declare a dummy subclass of same class to distinguish them went type checking. 

Like 
```python 
Kilogram = NewType("Kilogram", float)
Pound = NewType("Pound", float)

p = Pound(3)

def some_fun(a: Kilogram): 
    pass

some_fun(p)  # mypy have error here
```

Where `TypeVar` is a little more powerful type alias. When in above situation mypy will not point a problem. But You still have more meaningful names. 

If we would like to have option to validate usage using mypy then most probably for we will end with something like:

```python
ImageData_np = NewType("ImageData", np.ndarray)
ImageData_da = NewType("ImageData", da.Array)
...

ImageData = Union[ImageData_np,  ImageData_da, ...]
```

Reply:
I like that last option tbh!

Reply:
I convert to this approach Image layer. At this moment, I think that for all other layers, we require `np.ndarray`.

there is some problems with mypy and zarr that I do not have time to solve at this moment.

Reply:
Ok. The problem is the lack of annotation in zarr package... 

https://github.com/python/mypy/issues/6701

In this case mypy is to agresive

Reply:
Usage of union requires https://github.com/pyapp-kit/magicgui/pull/546